homeassistant/components/tuya/climate.py

```diff
@@ -140,170 +136,116 @@ def __init__(  # noqa: C901
-                device.status_range[self._current_temperature_dpcode].values
-            )
+            self._attr_max_temp = self._set_temperature.max_scaled
+            self._attr_min_temp = self._set_temperature.min_scaled
+            self._attr_target_temperature_step = self._set_temperature.step_scaled
 
         # Determine HVAC modes
         self._attr_hvac_modes = []
         self._hvac_to_tuya = {}
-        if DPCode.MODE in device.function:
-            data_type = EnumTypeData.from_json(device.function[DPCode.MODE].values)
+        if enum_type := self.get_enum_type(DPCode.MODE, function=True):
             self._attr_hvac_modes = [HVAC_MODE_OFF]
             for tuya_mode, ha_mode in TUYA_HVAC_TO_HA.items():
-                if tuya_mode in data_type.range:
+                if tuya_mode in enum_type.range:
                     self._hvac_to_tuya[ha_mode] = tuya_mode
                     self._attr_hvac_modes.append(ha_mode)
-        elif DPCode.SWITCH in device.function:
+        elif self.find_dpcode(DPCode.SWITCH, function=True):
             self._attr_hvac_modes = [
                 HVAC_MODE_OFF,
                 description.switch_only_hvac_mode,
             ]
 
         # Determine dpcode to use for setting the humidity
-        if DPCode.HUMIDITY_SET in device.function:
+        if int_type := self.get_integer_type(DPCode.HUMIDITY_SET, function=True):
             self._attr_supported_features |= SUPPORT_TARGET_HUMIDITY
-            self._set_humidity_dpcode = DPCode.HUMIDITY_SET
-            type_data = IntegerTypeData.from_json(
-                device.function[DPCode.HUMIDITY_SET].values
-            )
-            self._set_humidity_type = type_data
-            self._attr_min_humidity = int(type_data.min_scaled)
-            self._attr_max_humidity = int(type_data.max_scaled)
+            self._set_humidity = int_type
+            self._attr_min_humidity = int(int_type.min_scaled)
+            self._attr_max_humidity = int(int_type.max_scaled)
 
         # Determine dpcode to use for getting the current humidity
-        if (
-            DPCode.HUMIDITY_CURRENT in device.status
-            and DPCode.HUMIDITY_CURRENT in device.status_range
-        ):
-            self._current_humidity_dpcode = DPCode.HUMIDITY_CURRENT
-            self._current_humidity_type = IntegerTypeData.from_json(
-                device.status_range[DPCode.HUMIDITY_CURRENT].values
-            )
-
-        # Determine dpcode to use for getting the current humidity
-        if (
-            DPCode.HUMIDITY_CURRENT in device.status
-            and DPCode.HUMIDITY_CURRENT in device.status_range
-        ):
-            self._current_humidity_dpcode = DPCode.HUMIDITY_CURRENT
-            self._current_humidity_type = IntegerTypeData.from_json(
-                device.status_range[DPCode.HUMIDITY_CURRENT].values
-            )
+        self._current_humidity = self.get_integer_type(DPCode.HUMIDITY_CURRENT)
```

Comment:
Are dpcodes always matched to the same type? Right now the type logic is stored in the class and another cleanup could be attached to the dpcode so it’s defined once.

Maybe it’s possible  to have a list of supported dpcodes that the base class maps to types and it’s data automatically

Reply:
> Are dpcodes always matched to the same type? 

They are not, which is part of the problem.

> Right now the type logic is stored in the class and another cleanup could be attached to the dpcode so it’s defined once.

Depending on how this works out and solves things (hopefully), I would love to just build on top of the upstream library (or extend on it) to provide these data types natively/directly. Including being able to just set the values directly onto these objects (without having to handle them again locally).

So something like (pseudo):

```py
datapoint = self.find_datapoint(DPCode.TEMP)
state = datapoint.scaled_value
await datapoint.set_scaled_value(21.0)
```

While this PR is not close to that, it does move towards more central/less logic.

My focus right now is mainly the fallback. Implementing the fallback on all the places would have been hideous to do.

Reply:
yes that looks like a great API.